ext/standard/proc_open.c

```diff
@@ -507,6 +799,13 @@ PHP_FUNCTION(proc_open)
 #endif
 	php_process_id_t child;
 	struct php_process_handle *proc;
+
+  /* zend_strings which may be allocated while processing the descriptorspec
+   * we need to make sure that each one allocated is released once and only once */
+	zend_string *ztype = NULL, *zmode = NULL, *zfile = NULL;
+#define cleanup_zend_string(ptr) do { if (ptr != NULL) { zend_string_release(ptr); ptr = NULL; } } while(0)
```

Comment:
This is `zend_tmp_string_release()`.

Reply:
OK. Though the macros are still needed.

The point of these macros is that the function is loaded with `goto exit_fail;`s and it would be  cumbersome to ensure that the correct `zend_strings` are released at each such point. So when releasing them, we set the pointer to `NULL` so there's no danger of freeing them twice.

Reply:
I see. I don't really like this, let's leave it for now.

Reply:
In the interest of reducing non-local state, I've moved the code that uses these strings into a separate function in https://github.com/php/php-src/commit/2dc4481fa90319be2cb573166b3b1b9bbede3c35.